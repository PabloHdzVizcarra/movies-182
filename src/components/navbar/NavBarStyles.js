import styled from "styled-components";
import { respondTo } from "../../styles/_respondTo";

export const ContainNav = styled.div`
  display: grid;
  background-color: rgba(82, 72, 156, 1);
  grid-template-columns: 100%;
  grid-template-rows: auto auto;
  gap: 10px;
  align-items: center;



  ${respondTo.sm`
    padding: 0;
    display: flex;
    justify-content: space-between;
    align-items: center;
    gap: 0;
  `}

  ${respondTo.md`
    justify-content: space-between;
    display: flex;
    align-items: center;
  `}

  ${respondTo.lg`
    display: flex;
    align-items: center;
    justify-content: space-between;
  `}
`;

export const MainTitle = styled.div`
  font-size: 1.3rem;
  margin-left: 10px;
  font-weight: bold;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px;
  

  h1 {
    margin: 2px;
    cursor: pointer;
    font-family: 'Josefin Sans', sans-serif;
    color: white;
  }

  ${respondTo.sm`

    h1 {
      font-size: 28px;
    }
  `}

  ${respondTo.md`
    margin-top: -10px;
    justify-content: space-between;
    display: flex;
    align-items: center;

    h1 {
      margin-top: 18px;
      font-size: 44px;
    }
  `}

  ${respondTo.lg`
    margin-top: -10px;

    h1 {
      margin-top: 18px;
      font-size: 44px;
    }
  `}

`;

export const ContainButtons = styled.div`

  padding-bottom: 10px;
  display: grid;
  gap: 5px;
  

  ${respondTo.xs`
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;

    button {
      margin-top: 15px;
    }
  `}

  ${respondTo.sm`
    display: flex;
    align-items: center;
  `}
  ${respondTo.md`
    display: flex;
    align-items: center;
  `}
  ${respondTo.lg`
    display: flex;
    align-items: center;
  `}

  a {
    background-color: transparent;
    padding: 5px;
    border-radius: 6px;
    text-decoration: none;
    color: #FFFFFF;
    margin: 0 5px 0 5px;

    transition: 300ms ease background-color;
    transition: 300ms ease color;
    &:hover {
      background-color: transparent;
      color: #d9d9d9;
    }
  }

  svg {
    margin-left: 5px;

    transition: 300ms ease color;
    &:hover {
      color: #3385ff;
    }
  }
  
  button {
    background-color: transparent;
    border-radius: 6px;
    border: none;
    color: white;
    cursor: pointer;
    font-size: 15px;
    margin: 0 5px 0 5px;
    padding: 7px;
    text-align: inherit;
    text-decoration: none;


    transition: 300ms ease background-color;
    transition: 300ms ease color;
    &:hover {
      background-color: transparent;
      color: #ff4d4d;
    }
  }

  
`;
