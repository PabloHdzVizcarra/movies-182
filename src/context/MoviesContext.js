import React, { createContext, useReducer, useCallback, useState } from 'react';
import { moviesReducer } from '../reducers/moviesReducer';
import { types } from '../types/types';
import { db, firebase } from '../libs/firebase';

export const MoviesContext = createContext();



export const MoviesProvider = ({ children }) => {

  const [state, dispatch] = useReducer(moviesReducer, {
    movies: [],
    favoriteMovies: [],
    searchMovie: [],
    movie: {},
    loadingMovies: true
  });

  const [errorForm, setErrorForm] = useState({
    error: false,
    message: ''
  });
  
  const { movies, favoriteMovies, searchMovie, movie, loadingMovies } = state;
  

  const getTopRatedMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/top_rated?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingTopRatedMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getPopularMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingPopularMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getUpcomingMovies = useCallback(async () => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/upcoming?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingUpcomingMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getMoviesFirebase = async(uid) => {

    try {
      
      let allFavoritesMovies = [];

      await db.collection(uid)
      .doc("movies")
      .collection("favorites")
      .get()
      .then((querySnapshot) => {
        querySnapshot.forEach((doc) => {
          let movie;
          movie = { ...doc.data() }
          movie.docID = doc.id;
          allFavoritesMovies.push(movie)
        });
      });

      // const reversedArray = allFavoritesMovies.reverse();

      dispatch({
        type: types.getFavoritesMoviesFirebase,
        payload: allFavoritesMovies
      })

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieSearch = async (key) => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/search/movie?api_key=${apiKey}&query=${key}`);
      
      const { results } = await data.json();
      console.log(results);

      if (results.length === 0) {
        return setErrorForm({
          error: true,
          message: 'Intenta buscando otra pelicula'
        })
      }

      setErrorForm({
        error: false,
        message: ''
      })

      dispatch({
        type: types.searchMovies,
        payload: results
      });

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieID = async (movieID) => {

    try {

      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/movie/${movieID}?api_key=${apiKey}&language=es-ES`);

      const resp = await data.json();
      
      dispatch({
        type: types.selectedMovie,
        payload: resp
      })
      
    } catch (error) {
      console.log(error);
    }
  }



  const addFavoriteMovie = async (movie) => {

    const {
      poster_path,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    } = movie;

    const imgMovie = `https://image.tmdb.org/t/p/w500/${poster_path}`;

    const favoriteMovie = {
      imgMovie,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    }

    const { uid } = firebase.auth().currentUser;

    if (uid) {

      const { id } = await db.collection(uid).doc('movies').collection('favorites').add(favoriteMovie);

      favoriteMovie.firebaseID = id;

      dispatch({
        type: types.addFavoritesMovies,
        payload: favoriteMovie
      })
      
      console.log('PELICULA AGREGADA A FAVORITOS');

    } else {
      throw new Error('No hay usuario autenticado')
    }

  }

  const deleteMovieFromFavorites = async (userID, docID) => {
    console.log(docID);

    dispatch({
      type: types.removeMovieFromFavorites,
      payload: docID
    })
    

    await db.collection(userID).doc('movies').collection('favorites').doc(docID)
      .delete();

    return console.log(`PELICULA ELIMINADA`);
  }
  
  return (
    <MoviesContext.Provider value={{
      movie,
      movies,
      favoriteMovies,
      searchMovie,
      loadingMovies,
      errorForm,
      getMovieID,
      getPopularMovies,
      getTopRatedMovies,
      getUpcomingMovies,
      getMovieSearch,
      addFavoriteMovie,
      getMoviesFirebase,
      deleteMovieFromFavorites,
      dispatch
    }}>
      {children}
    </MoviesContext.Provider>
  )
}