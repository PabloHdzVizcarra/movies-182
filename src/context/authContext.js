import React from 'react'
import { types } from '../types/types';

const AuthStateContext = React.createContext();
const AuthDispatchContext = React.createContext();

function authReducer(state, action) {
  switch (action.type) {

    case types.setActiveUser:
      return {
        ...state,
        activeUser: action.payload,
        isActived: true
      }

    case types.logoutUser:
      return {
        ...state,
        activeUser: {},
        isActived: false
      }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}


function AuthProvider({ children }) {
  const [state, dispatch] = React.useReducer(authReducer, {
    activeUser: {},
    isActived: false
  });

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

function useAuthState() {
  const context = React.useContext(AuthStateContext);

  if (context === undefined) {
    throw new Error('useAuthState must be used within AuthProvider')
  }

  return context
}

function useAuthDispatch() {
  const context = React.useContext(AuthDispatchContext);

  if (context === undefined) {
    throw new Error('useAuthDispatch must be used within AuthProvider')
  }

  return context
}

export {
  AuthProvider,
  useAuthState,
  useAuthDispatch
}