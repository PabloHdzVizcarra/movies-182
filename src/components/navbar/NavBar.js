import React, { useContext } from "react";
import { Link, NavLink, useHistory } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { ContainNav, MainTitle, ContainButtons } from "./NavBarStyles";
import { firebase, googleAuthProvider } from "../../libs/firebase";
import { useAuthState, useAuthDispatch } from "../../context/authContext";
import { types } from "../../types/types";
import { MoviesContext } from "../../context/MoviesContext";

export const NavBar = () => {
  const { activeUser, isActived } = useAuthState();
  const dispatch = useAuthDispatch();
  const { dispatch: dispatchMovies } = useContext(MoviesContext);
  const history = useHistory();

  const handleClickLoginGoogle = () => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(function (result) {})
      .catch(function (error) {
        console.log(error);
      });
  };

  const handleClickLogout = () => {
    firebase.auth().signOut();
    dispatch({
      type: types.logoutUser,
    });
  };

  const cleanFavoriteMovies = () => {
    dispatchMovies({
      type: types.cleanMovies,
    });
  };

  const handlePushPage = () => {
    history.push("/");
  };

  return (
    <ContainNav>
      <MainTitle>
        <h1 onClick={handlePushPage}>Fun Movies</h1>
      </MainTitle>
      <ContainButtons>
        <NavLink
          activeStyle={{
            color: "#d9d9d9",
            pointerEvents: "none",
            cursor: "default",
          }}
          onClick={cleanFavoriteMovies}
          to="/popular"
        >
          Populares
        </NavLink>
        <NavLink
          activeStyle={{
            color: "#d9d9d9",
            pointerEvents: "none",
            cursor: "default",
          }}
          onClick={cleanFavoriteMovies}
          to="/top"
        >
          Mejor valoradas
        </NavLink>
        <NavLink
          activeStyle={{
            color: "#d9d9d9",
            pointerEvents: "none",
            cursor: "default",
          }}
          onClick={cleanFavoriteMovies}
          to="/upcoming"
        >
          Proximas
        </NavLink>
        <NavLink
          activeStyle={{
            color: "#d9d9d9",
            pointerEvents: "none",
            cursor: "default",
          }}
          onClick={cleanFavoriteMovies}
          to="/search"
        >
          Buscar
        </NavLink>

        {isActived && (
          <NavLink
            to="/favorites"
            activeStyle={{
              color: "#d9d9d9",
              pointerEvents: "none",
              cursor: "default",
            }}
          >
            Favoritas
          </NavLink>
        )}

        {Object.keys(activeUser).length === 0 ? (
          <Link to="/">
            Inicia Sesion
            <FontAwesomeIcon
              icon={["fab", "google"]}
              onClick={handleClickLoginGoogle}
            />
          </Link>
        ) : (
          <button onClick={handleClickLogout}>Cerrar sesion</button>
        )}
      </ContainButtons>
    </ContainNav>
  );
};
