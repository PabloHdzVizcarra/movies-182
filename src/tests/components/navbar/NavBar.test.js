import "@testing-library/jest-dom";
import React from "react";
import { mount } from "enzyme";
import { AuthProvider, AuthStateContext } from "../../../context/authContext";
import { MoviesContext, MoviesProvider } from "../../../context/MoviesContext";
import { MemoryRouter } from "react-router-dom";
import { NavBar } from "../../../components/navbar/NavBar";
import { render, screen, fireEvent } from "@testing-library/react";

describe("test in <NavBar /> component", () => {
  //Enzyme
  const contextValue = {
    deleteMovieFromFavorites: jest.fn(),
  };

  const wrapper = mount(
    <AuthProvider>
      <MoviesContext.Provider value={contextValue}>
        <MemoryRouter>
          <NavBar />
        </MemoryRouter>
      </MoviesContext.Provider>
    </AuthProvider>
  );

  // react testing library

  const Wrapper = ({ children }) => {
    return (
      <AuthProvider>
        <MoviesProvider>
          <MemoryRouter>{children}</MemoryRouter>
        </MoviesProvider>
      </AuthProvider>
    );
  };

  const customRender = (ui, options) => {
    render(ui, { wrapper: Wrapper, ...options });
  };

  test("the component must be correctly rendered ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("the login button should work", () => {
    customRender(<NavBar />);
    expect(screen.getByText(/Inicia Sesion/i)).toBeInTheDocument();
  });

  test('when the user is logged in, the "Cerrar sesion" button should appear', async () => {
    const Wrapper = ({ children }) => {
      return (
        <AuthProvider>
          <AuthStateContext.Provider
            value={{
              isActived: true,
              activeUser: {
                displayName: "Pablo Hernandez",
              },
            }}
          >
            <MoviesContext.Provider
              value={{
                dispatch: jest.fn(),
              }}
            >
              <MemoryRouter>{children}</MemoryRouter>
            </MoviesContext.Provider>
          </AuthStateContext.Provider>
        </AuthProvider>
      );
    };

    const customRender = (ui, options) => {
      render(ui, { wrapper: Wrapper, ...options });
    };

    customRender(<NavBar />);
    expect(screen.getByText(/Cerrar sesion/i)).toBeInTheDocument();
  });

  test('when you click on "Cerrar sesion", the users current session should be closed and the text "Cerrar sesion" should disappear', async () => {
    const Wrapper = ({ children }) => {
      return (
        <AuthProvider>
          <AuthStateContext.Provider
            value={{
              isActived: true,
              activeUser: {
                displayName: "Pablo Hernandez",
              },
            }}
          >
            <MoviesContext.Provider
              value={{
                dispatch: jest.fn(),
              }}
            >
              <MemoryRouter>{children}</MemoryRouter>
            </MoviesContext.Provider>
          </AuthStateContext.Provider>
        </AuthProvider>
      );
    };

    const customRender = (ui, options) => {
      render(ui, { wrapper: Wrapper, ...options });
    };

    customRender(<NavBar />);

    const node = screen.getByText(/Cerrar sesion/i);
    fireEvent.click(node);
  });
});
