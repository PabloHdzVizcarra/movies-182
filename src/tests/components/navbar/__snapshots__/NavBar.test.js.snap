// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <NavBar /> component the component must be correctly rendered  1`] = `
<div
  className="sc-AxjAm fvhgcF"
>
  <div
    className="sc-AxirZ dGcRPv"
  >
    <h1
      onClick={[Function]}
    >
      PelisData
    </h1>
  </div>
  <div
    className="sc-AxiKw fbYlOk"
  >
    <a
      aria-current={null}
      href="/popular"
      onClick={[Function]}
    >
      Populares
    </a>
    <a
      aria-current={null}
      href="/top"
      onClick={[Function]}
    >
      Mejor valoradas
    </a>
    <a
      aria-current={null}
      href="/upcoming"
      onClick={[Function]}
    >
      Proximas
    </a>
    <a
      aria-current={null}
      href="/search"
      onClick={[Function]}
    >
      Buscar
    </a>
    <a
      href="/"
      onClick={[Function]}
    >
      Inicia Sesion
      <i
        className="fa"
      />
    </a>
  </div>
</div>
`;
